gcp/modules/gcp-stackdriver-alerting/client.rb

```diff
@@ -0,0 +1,172 @@
+require "json"
+require "google/cloud/monitoring"
+
+@project_id = ENV['PROJECT_ID']
+@destroy_triggered = ENV['DESTROY']
+@debug_mode = ENV['DEBUG']
+@resource_dir="#{File.expand_path(File.dirname(__FILE__))}/resources_rendered"
+
+def get_notification_channel_attribute(notification_channel)
+  if notification_channel["labels"]["email_address"]
+    result = notification_channel["labels"]["email_address"]
+  end
+
+  return result
+end
+
+def get_uptime_check_attribute(uptime_check)
+  if uptime_check["display_name"]
+    result = uptime_check["display_name"]
+  end
+
+  return result
+end
+
+def get_alert_policy_attribute(alert_policy)
+  if alert_policy["display_name"]
+    result = alert_policy["display_name"]
+  end
+
+  return result
+end
+
+def process_notification_channels(notification_channels = [])
+  notification_channel_service_client = Google::Cloud::Monitoring::NotificationChannel.new(version: :v3)
+  formatted_parent = Google::Cloud::Monitoring::V3::NotificationChannelServiceClient.project_path(@project_id)
+
+  stackdriver_notification_channels = {}
+  processed_notification_channels = {}
+
+  notification_channel_service_client.list_notification_channels(formatted_parent).each do |notification_channel|
+    puts notification_channel.to_json if @debug_mode
+    stackdriver_notification_channels[get_notification_channel_attribute(notification_channel)] = notification_channel
+  end
+
+  notification_channels.each do |notification_channel|
+    notification_channel_attribute = get_notification_channel_attribute(notification_channel)
+
+    if stackdriver_notification_channels[notification_channel_attribute]
+      notification_channel["name"] = stackdriver_notification_channels[notification_channel_attribute]["name"]
+      puts "Updating notification channel \"#{notification_channel_attribute}\"..."
+      notification_channel_service_client.update_notification_channel(notification_channel)
+    else
+      puts "Creating notification channel \"#{notification_channel_attribute}\"..."
+      notification_channel = notification_channel_service_client.create_notification_channel(formatted_parent, notification_channel)
+    end
+
+    processed_notification_channels[notification_channel_attribute] = notification_channel["name"]
+  end
+
+  stackdriver_notification_channels.each do |name, notification_channel|
+    notification_channel_attribute = get_notification_channel_attribute(notification_channel)
+
+    unless processed_notification_channels.include? notification_channel_attribute or @debug_mode
```

Comment:
It is the other way around – enabling debug mode disables deletion of Stackdriver resources.

Reply:
I still find this behavior pretty unintuitive. (I also find `unless X or Y` hard to reason about but maybe that's just me.)

If skipping deletion in debug mode is desirable, I might do it like this:
```
unless X
  if @debug_mode
    puts "[DEBUG] Skipping deletion of #{name} because @debug_mode is set"
  else
    do_the_thing
  end
end
```

Not a big deal, ok to skip if you think this is unlikely to bite anyone.